### PR TITLE
feat: use new llvm gc api

### LINKED ISSUE #362: GC Root trace optimization
- use statepoint api for gc
- manually walk the stack instead of using libunwind

### PR DESCRIPTION
**Pull request by Chronostasys**
change llvm.gcroot to statepoint api

TODOs:
- [x] global variable mark
- [x] code clean up
- [x] docs
- [x] debug

close #362

### CODE REVIEW

**Comment by xieyuschen** [comment]
can rebase the commits to split them into several commits with meaningful commit info. It's too large hence squash them together is not a good idea.

File: immix/llvm/memory_manager.cpp [hunk 1]
```diff
@@ -37,9 +35,13 @@ static uint8_t *roundTripAllocateDataSection(void *object, uintptr_t size,
 {
   auto alloc_addr = static_cast<SectionMemoryManager *>(object)->allocateDataSection(
       size, alignment, sectionID, sectionName, isReadOnly);
-  // printf("sec name: %s\n", sectionName);
-  auto mapName = "_GC_MAP_";
-  if (strncmp(sectionName, mapName, strlen(mapName)) == 0)
+// printf("sec name: %s\n", sectionName);
+#ifdef __APPLE__
```
Comment (xieyuschen) [inline immix/llvm/memory_manager.cpp:39]: Does the `__APPLE__` macro represent the MacOs? Can document why we need different `llvmSectionName` to fit the OS.

File: immix/llvm/memory_manager.cpp [hunk 2]
```diff
@@ -80,12 +82,16 @@ static void roundTripDestroy(void *object)
...
   delete static_cast<SectionMemoryManager *>(object);
 }
+
+
 extern "C"
 {
-  void CreatePLJITEngine(LLVMExecutionEngineRef *jit, LLVMModuleRef module, unsigned int opt, stackmap_cb cb)
+  LLVMExecutionEngineRef * CreatePLJITEngine( LLVMModuleRef module, unsigned int opt, stackmap_cb cb)
   {
+    LLVMExecutionEngineRef *jit = new LLVMExecutionEngineRef();
```
Comment (xieyuschen) [inline immix/llvm/memory_manager.cpp:91]: I'm curious when do you free this pointer as you `new` manually. Can consider to use `shared_ptr` or `unique_ptr.

File: immix/llvm/plimmix.cpp [hunk 3]
```diff
@@ -32,6 +35,22 @@ namespace
         X("plimmix", "pivot-lang immix garbage collector.");
 }
 
+extern "C" LLVMMemoryBufferRef parse_ir(char* ir)
+{
+    SMDiagnostic Err;
+    auto Ctx = std::make_unique<LLVMContext>();
+    auto sIr = std::string(ir);
+    auto Mod = parseIRFile(sIr, Err, *Ctx);
```
Comment (xieyuschen) [inline immix/llvm/plimmix.cpp:43]: can consider to use snake case to make the names of variables consistent.

File: immix/llvm/plimmix_pass.cpp [hunk 4]
```diff
@@ -49,31 +49,54 @@ namespace
   */
   void immixPassLogic(Module &M)
   {
+    auto isMain = false;
     for (auto FB = M.functions().begin(), FE = M.functions().end(); FB != FE; ++FB)
```
Comment (xieyuschen) [inline immix/llvm/plimmix_pass.cpp:53]: it's more common to write this, IIRC. 
```suggestion
    for (auto FB = M.functions().begin(); FB != FE = M.functions().end(); FB++)
```

File: immix/llvm/plimmix_pass.cpp [hunk 5]
```diff
@@ -49,31 +49,54 @@ namespace
...
-    assertArgs.push_back(g);
-    Function *gc_init_f;
-    std::tie(gc_init_f, std::ignore) = createSanitizerCtorAndInitFunctions(M, "__gc_init_stackmap", "immix_gc_init", argTypes, assertArgs);
-    gc_init_f->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
-    // appendToCompilerUsed(M, gc_init_f);
-    appendToGlobalCtors(M, gc_init_f, 1000);
+    if (isMain)
+    {
+      // auto gc_init_c = M.getOrInsertFunction("__gc_init_stackmap", Type::getVoidTy(M.getContext()));
+      auto immix_init_c = M.getOrInsertFunction("immix_gc_init", Type::getVoidTy(M.getContext()), PointerType::get(IntegerType::get(M.getContext(), 8), 0));
```
Comment (xieyuschen) [inline immix/llvm/plimmix_pass.cpp:69]: could you add some comments to explain the procedures here? As you changed almost the whole function implementation but haven't changed the function comment.

File: immix/src/consts.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ pub const EVA_BLOCK_PROPORTION: f64 = 0.025;
 
 pub const ALIGN: usize = 4096;
 
-pub const THRESHOLD_PROPORTION: f64 = 1.75;
+pub const THRESHOLD_PROPORTION: f64 = 1000.75;
```
Comment (xieyuschen) [inline immix/src/consts.rs:15]: can explain why the value is `1000.75` instead of something else.

File: immix/src/llvm_stackmap/mod.rs [hunk 7]
```diff
@@ -1,33 +1,22 @@
...
-
+#![allow(dead_code)]
 use int_enum::IntEnum;
 use rustc_hash::FxHashMap;
 
 use crate::ObjectType;
+use std::marker::PhantomData;
+
+const SAFE_POINT_ID: u64 = 2882400000;
+// const GC_ROOT_ID: u64 = 114514;
```
Comment (xieyuschen) [inline immix/src/llvm_stackmap/mod.rs:9]: remove useless code?

File: immix/src/llvm_stackmap/mod.rs [hunk 8]
```diff
@@ -47,46 +36,16 @@ fn get_first_function_addr(mapptr: *const u8) -> *const u8 {
 
 #[derive(Debug, Clone)]
 pub struct Function {
-    pub addr: *const u8,
-    pub root_size: i32,
+    // pub addr: *const u8,
+    // pub offset:u32,
+    // pub root_size: i32,
```
Comment (xieyuschen) [inline immix/src/llvm_stackmap/mod.rs:41]: ditto

File: src/ast/builder/llvmbuilder.rs [hunk 9]
```diff
@@ -202,13 +201,95 @@ impl<'a, 'ctx> LLVMBuilder<'a, 'ctx> {
...
             block_table: Arc::new(RefCell::new(FxHashMap::default())),
             block_reverse_table: Arc::new(RefCell::new(FxHashMap::default())),
-            heap_stack_map: Arc::new(RefCell::new(FxHashMap::default())),
             optimized: Arc::new(RefCell::new(false)),
             used: Default::default(),
             difile: Cell::new(diunit.get_file()),
             optlevel: opt,
         }
     }
+    fn alloc_with_f(
```
Comment (xieyuschen) [inline src/ast/builder/llvmbuilder.rs:210]: add a comment please

File: src/ast/builder/llvmbuilder.rs [hunk 10]
```diff
@@ -254,30 +335,26 @@ impl<'a, 'ctx> LLVMBuilder<'a, 'ctx> {
...
         if let Some(p) = declare {
             self.build_dbg_location(p);
-            self.insert_var_declare(
-                name,
-                p,
-                &PLType::Pointer(Arc::new(RefCell::new(pltype.clone()))),
-                self.get_llvm_value_handle(&stack_root.as_any_value_enum()),
-                ctx,
-            );
+            // TODO
```
Comment (xieyuschen) [inline src/ast/builder/llvmbuilder.rs:338]: todo what? does the commented code below is the item of todo?

File: src/ast/builder/no_op_builder.rs [hunk 11]
```diff
@@ -515,9 +521,9 @@ impl<'a, 'ctx> IRBuilder<'a, 'ctx> for NoOpBuilder<'a, 'ctx> {
 
     unsafe fn store_with_aoto_cast(&self, _ptr: ValueHandle, _value: ValueHandle) {}
 
-    fn stack_alloc(&self, _name: &str, _ctx: &mut Ctx<'a>, _tp: &PLType) -> ValueHandle {
-        0
-    }
+    // fn stack_alloc(&self, _name: &str, _ctx: &mut Ctx<'a>, _tp: &PLType) -> ValueHandle {
```
Comment (xieyuschen) [inline src/ast/builder/no_op_builder.rs:524]: ditto

File: src/ast/compiler.rs [hunk 12]
```diff
@@ -147,18 +162,25 @@ pub fn pl_link(llvmmod: Module, oxbjs: Vec<PathBuf>, out: String, op: Options) {
 
     out.push_str(".bc");
 
-    let tm = crate::ast::builder::llvmbuilder::get_target_machine(op.optimization.to_llvm());
-    llvmmod.set_triple(&tm.get_triple());
-    llvmmod.set_data_layout(&tm.get_target_data().get_data_layout());
+    // let tm = crate::ast::builder::llvmbuilder::get_target_machine(op.optimization.to_llvm());
```
Comment (xieyuschen) [inline src/ast/compiler.rs:165]: ditto

File: src/ast/ctx/builtins.rs [hunk 13]
```diff
@@ -474,13 +484,126 @@ fn emit_arr_copy<'a, 'b>(
...
+            let len_raw =
+                ctx.try_load2var(Default::default(), len_raw, builder, &PLType::new_i64())?;
             builder.build_memcpy(from_raw, &a1.element_type.borrow(), to_raw, len_raw, ctx);
             Ok(Default::default())
         }
         _ => unreachable!(),
     }
 }
 
+fn emit_arr_slice<'a, 'b>(
```
Comment (xieyuschen) [inline src/ast/ctx/builtins.rs:496]: add a comment for this function.

File: src/ast/node/function/generator.rs [hunk 14]
```diff
@@ -124,13 +124,13 @@ pub(crate) fn end_generator<'a>(
...
         .unwrap();
     unsafe { builder.store_with_aoto_cast(ptr, funcvalue) };
-    let ret_load = builder.build_load(
-        data.borrow().ret_handle,
-        "ret_load",
-        &b.ret_type.as_ref().unwrap().borrow(),
-        child,
-    );
-    builder.build_return(Some(ret_load));
+    // let ret_load = builder.build_load(
```
Comment (xieyuschen) [inline src/ast/node/function/generator.rs:127]: ditto

File: vm/src/mutex/mod.rs [hunk 15]
```diff
@@ -6,6 +6,14 @@ use std::{
 
 use internal_macro::is_runtime;
 
+#[cfg(feature = "jit")]
+pub fn reg() {
```
Comment (xieyuschen) [inline vm/src/mutex/mod.rs:10]: add a comment please

File: immix/llvm/memory_manager.cpp (hunk 2 above)
Comment (Chronostasys) [reply immix/llvm/memory_manager.cpp:91]: in fact, LLVMExecutionEngineRef itself is a pointer, so there's no need to initialize it. I'll remove new expression later

File: immix/llvm/plimmix_pass.cpp (hunk 4 above)
Comment (Chronostasys) [reply immix/llvm/plimmix_pass.cpp:53]: It's copied from LLVM source code... Please do not submit suggestions like this again if it's not severe.

File: immix/src/consts.rs (hunk 6 above)
Comment (Chronostasys) [reply immix/src/consts.rs:15]: 这个其实就是感觉1.75太小了，这个参数影响gc的回收频率，以后可以调研一下什么数字合适

File: immix/src/llvm_stackmap/mod.rs (hunk 8 above)
Comment (Chronostasys) [reply immix/src/llvm_stackmap/mod.rs:24]: Although those code were commented for performance reason, leave them here could help debugging if there's bug in stackmap reader.

**Pull request merged**